### PR TITLE
Use caching objects from cachem package

### PR DESCRIPTION
**Pull request by wch**
This PR updates memoise to use the caching objects from the [cachem](https://github.com/r-lib/cachem) package. These caching objects support automatic pruning, so that they won't grow indefinitely.

### CODE REVIEW

**Comment by wch** [comment]
I believe the covr tests are failing because covr manipulates the AST of the code, but this code expects things to be in a specific place:

https://github.com/wch/memoise/blob/b1c628d5e9cd7fcac957bca918c516a2d7324573/R/memoise.R#L185-L195

The `has_cache()` and `drop_cache()` functions do something similar, but they don't result in problems with covr.

**Comment by jimhester** [comment]
I don't think you should try to handle old style caches at all, people will just need to generate new caches.

File: R/memoise.R [hunk 1]
```diff
@@ -142,14 +150,18 @@ memoise <- memoize <- function(f, ..., envir = environment(f), cache = cache_mem
     args <- c(lapply(called_args, eval, parent.frame()),
               lapply(default_args, eval, envir = environment()))
 
-    hash <- encl$`_cache`$digest(
-      c(as.character(body(encl$`_f`)), args,
-        lapply(encl$`_additional`, function(x) eval(x[[2L]], environment(x))))
+    # Use getNamespace() instead of `::`, because the latter is slow.
```
Comment (jimhester) [inline R/memoise.R:153]: Do you have a benchmark to support this? I struggle to see when this difference would ever amount to a significant amount in a real application.

You may observe some overhead with a trivial microbenchmark (I added some possibly faster alternatives to your code as well)

``` r
ns <- getNamespace("base")
bench::mark(base::identity(1),
  getNamespace("base")$identity(1),
  getNamespace("base")[["identity"]](1),
  ns[["identity"]](1),
  .subset2(ns, "identity")(1)
)
#> # A tibble: 5 x 6
#>   expression                                 min   median `itr/sec` mem_alloc
#>   <bch:expr>                            <bch:tm> <bch:tm>     <dbl> <bch:byt>
#> 1 base::identity(1)                       3.95µs   4.49µs   200947.        0B
#> 2 getNamespace("base")$identity(1)         778ns    962ns   918720.        0B
#> 3 getNamespace("base")[["identity"]](1)    707ns    840ns  1054828.        0B
#> 4 ns[["identity"]](1)                      334ns    421ns  2069693.        0B
#> 5 .subset2(ns, "identity")(1)              338ns    419ns  2083773.        0B
#> # … with 1 more variable: `gc/sec` <dbl>
```

But this difference mostly goes away soon as you use any real world example, e.g. here I digest the `digest::digest` function itself.

``` r
ns <- getNamespace("digest")
dg <- digest::digest
bench::mark(digest::digest(dg),
  getNamespace("digest")$digest(dg),
  getNamespace("digest")[["digest"]](dg),
  ns[["digest"]](dg),
  .subset2(ns, "digest")(dg)
)
#> # A tibble: 5 x 6
#>   expression                               min median `itr/sec` mem_alloc
#>   <bch:expr>                             <bch> <bch:>     <dbl> <bch:byt>
#> 1 digest::digest(dg)                     198µs  212µs     4376.    52.5KB
#> 2 getNamespace("digest")$digest(dg)      190µs  203µs     4612.    36.8KB
#> 3 getNamespace("digest")[["digest"]](dg) 190µs  202µs     4692.    36.8KB
#> 4 ns[["digest"]](dg)                     192µs  200µs     4748.    36.8KB
#> 5 .subset2(ns, "digest")(dg)             191µs  202µs     4696.    36.8KB
#> # … with 1 more variable: `gc/sec` <dbl>
```

<sup>Created on 2020-11-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Personally I think using this type of micro-optimization hurts maintainability with negligible improvements in performance.

Also the current implementation does not presume use of the digest package, whereas this implementation does. I think that is a significant backstep in genericity.

Comment (wch) [reply R/memoise.R:153]: I think that the inputs to a memoized function aren't necessarily large -- there are cases where the inputs are small but they result in a large computation. But I'm OK going back to `::` for maintainability. (I wish that something as basic as `::` didn't have as much overhead as it does.)

As for the genericity and the digest package, do you think that's an issue in practice? Are there other functions from other packages that one would want to use to transform the R object into a key? If so, we could allow the user to pass in a function to transform the R object, and have it default to `function(x) digest(x, algo = "spookyhash")`.

**Comment by wch** [comment]
As for supporting old-style caches, I wanted to make sure that if people had made their own cache objects, it wouldn't break for them.

**Comment by wch** [comment]
I've updated the PR so `memoise()` now takes an argument `hash`, which is a function that takes an R object as input and returns a hash of that object. The default value is:

```
  hash = function(x) digest::digest(x, algo = "spookyhash"))
```

Because the default is defined in the `memoise()` function, it captures the environment. I don't know if this is a problem or not. It looks like efforts were made to avoid capturing the function environment, by using `memo_f_env`, but it's possible that the purpose of that environment is not necessarily to avoid capturing the environment, but rather to make the memoized function have the original function's environment as an ancestor.

File: R/memoise.R [hunk 2]
```diff
@@ -49,12 +51,14 @@
...
 #' }
 #' @name memoise
-#' @title Memoise a function.
 #' @param f     Function of which to create a memoised copy.
 #' @param ... optional variables to use as additional restrictions on
 #'   caching, specified as one-sided formulas (no LHS). See Examples for usage.
 #' @param envir Environment of the returned function.
-#' @param cache Cache function.
+#' @param cache Cache object. The default is a [cachem::cache_mem()] with a max
+#'   size of 512 MB.
```
Comment (wch) [inline R/memoise.R:59]: Fix this number

**Review by jimhester — APPROVED** [review]

**Pull request merged**